*: add testutils as disallowed import to top-level packages

rafiss opened the pull request:
See the individual commits to make this easier to review.

This PR is a step towards completing https://github.com/cockroachdb/cockroach/issues/81375, with the ultimate goal
of no longer depending on test packages in our production binary. More
work is needed to avoid depending on any of the subpackages inside of `testutils`.

informs https://github.com/cockroachdb/cockroach/issues/81375

rafiss committed:
kv: move test_utils file to new kvtestutils package

Release note: None

rafiss committed:
sql: move testing function to sqltestutils package

Release note: None

rafiss committed:
sql: disallow import of testutils package

Release note: None

rafiss committed:
server: break dependency on testutils

Release note: None

rafiss committed:
spanconfigptsreader: break deps on testing packages

Release note: None

rafiss committed:
server: break dependency on testutils/sqlutils

Release note: None

rafiss committed:
server: mark testutils as a disallowed import

Release note: None

rafiss committed:
cli: break dependency on testutils

Release note: None

rafiss committed:
crosscluster/producer: break testutils dependency

Release note: None

rafiss committed:
crosscluster/logical: break testutils dependency

Release note: None

rafiss committed:
cmd/cockroach: add testutils to disallowed deps

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141442)
<!-- Reviewable:end -->

rickystewart commented:
Looks like there is a dependency cycle in `pkg/sql:sql_test`, I'll wait to review until that's all resolved.

rafiss commented:
@rickystewart this is ready for a review.

rickystewart reviewed (approved):
Looks good, a couple thoughts:

1. I recommend a brief summary (in the PR description, etc.) to describe what the impact of this change is, including why the refactor would be beneficial. Might be worth mentioning explicitly that this is a step on the way to completing #81375, but there are still dependencies have not been addressed.
2. I'm not sure if you plan on doing any more work in this area but setting `testonly = 1` on packages that are explicitly test-only is far and away the simplest and most robust tool we have to ensuring that code will not get linked into `cockroach` or any other top-level binaries. (I would have gone and set this for all packages under `pkg/testutils` years ago, but that was always impossible as these packages were never truly test-only.) I recommend being aggressive about adding it where possible. Its use is complimentary to `disallowed_imports_test` as `disallowed_imports_test` describes what a target is allowed to depend on, while `testonly = 1` has an effect on all packages that would depend on that target.

rickystewart commented on pkg/kv/kvtestutils/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+go_library(
+    name = "kvtestutils",
+    srcs = ["test_utils.go"],
+    importpath = "github.com/cockroachdb/cockroach/pkg/kv/kvtestutils",
```
(I haven't tested this a ton, let me know if this introduces build failures I wasn't aware of and it could be omitted.)

rafiss commented:
Thanks for the review. The thing that motivated this PR was to avoid pulling in the sideeye dependency, but I see that was addressed separately here: b92432b86ab4298ef13cd0f95a4b84972d9aa317. BTW, it seems like the new sideeye package you introduced there also would be a good candidate for `testonly = 1`.

I updated the PR description to mention the general goal of not depending on test packages.

Let me see what happens when I add `testonly = 1` to this.

rafiss committed:
kvtestutils: mark package as testonly

Release note: None

rickystewart commented:
Sorry for any confusion, in this case I was only specifically recommending `testonly = 1` for `pkg/kv/kvtestutils/BUILD.bazel` as that was the only one that seemed to plausibly allow it.

rickystewart commented:
I'm not sure if the CI failure seen in that `bors` run is going to be stable. It seems to be related to this PR, but I'm not sure.

rafiss committed:
sql: revert accidental test changes

84079c9db236a97dd59ebd6937b22a8ab464d5b1 accidentally changed some
testing code to stop testing with tenants and to make it use
sqlutils.PGUrl.

This reverts those accidental changes.

Release note: None

rafiss commented:
The issue was fixed. I needed to rebase on top of origin/master, and fix a usage of sqlutils.

bors r+
